README.md
# fff

Dart package for convenient work with color. It will allow you to convert
colors between rgb, rgba and hex formats. Also it can parse color from
strings specified formats and from list of color components or map.

## Usage

### A simple usage Color:

```dart
import 'package:fff/color.dart';

main() {

    // create black color in rgb format
    const black = const Color();

    // or
    const black = const Color(0, 0, 0);

    // output it
    print(black); // rgb(0, 0, 0)
    print(black.toHexString()); // 000000
    print(black.toRgbString()); // rgb(0, 0, 0)
    print(black.toRgbaString()); // rgba(0, 0, 0, 1.0)

    // create red color in rgba format
    const red = const Color(255, 0, 0, 1.0);
}
```

Output color components as List object

```dart
import 'package:fff/color.dart';

main() {

    // create red color
    var red = const Color(255);

    // output simple list
    print(red.toList()); // [255, 0, 0]

    // output by template
    print(
      red.toList(template: [
        Component.ALPHA,
        Component.BLUE,
        Component.GREEN,
        Component.RED
      ])
    );
    // [1.0, 0, 0, 255]

    // value of the alpha channel always output as double and range is
    // set from 0 to 1 rgb components are always output as int. But
    // this can be changed by setting the parameter to true asDouble
    print(red.toList(asDouble: true)); // [255.0, 0.0, 0.0]

    // change of range of color components
    print(red.toList(range: 1)); // [1, 0, 0]

    // You can specify the number of times the output. For example it
    // may be useful to specify the color of each vertex polygon model
    print(red.toList(rows: 2, range: 1));
    // [1.0, 0.0, 0.0, 1.0, 0.0, 0.0]
}
```

Output color components as Map object

```dart
import 'package:fff/color.dart';

main() {

    var temp = {
      "A": Component.RED,
      "B": Component.GREEN,
      "C": Component.BLUE
    };

    // are all exactly the same as in the first case, with the only
    // exception that here there is no parameter rows and default
    // template {"red": RED, "green": GREEN, "blue": BLUE, "alpha":
    // ALPHA}
    print(Color(255, 100, 50).toMap(template: temp, range: 1));
    // {A: 1.0, B: 0.39215686274509803, C: 0.19607843137254902}

}
```

In addition and subtraction of colors, the left is the main color. That is, in the end will be the color of the same size as the left color. For color in rgb format value of alpha is 1.0
 
 ```dart
print(Color(100, 0, 0) + Color(50, 0, 0)); // rgb(150, 0, 0)
print(Color(100, 0, 0) - Color(50, 0, 0)); // rgb(50, 0, 0)

print(Color(100, 0, 0) + Color(50, 0, 0, .1)); // rgb(150, 0, 0)
print(Color(100, 0, 0) - Color(50, 0, 0, .1)); // rgb(50, 0, 0)

print(Color(100, 0, 0, .5) + Color(50, 0, 0)); // rgba(150, 0, 0, 1.0)
print(Color(100, 0, 0, .5) - Color(50, 0, 0)); // rgba(50, 0, 0, 0.0)

print(Color(100, 0, 0, .5) + Color(50, 0, 0, .3)); // rgba(150, 0, 0, 0.8)
print(Color(100, 0, 0, .5) - Color(50, 0, 0, .3)); // rgba(50, 0, 0, 0.2)
```

### Color presets

Presets library contains 275 constants of colors from [Material design](http://www.google.com/design/spec/style/color.html)

```dart
import 'package:fff/presets.dart';

main() {
    print(red + green); // rgb(255, 242, 134)
}
```

### Parse color

```dart
import 'package:fff/parser.dart';

main() {

    print(ColorParser("#FFF")); // rgb(255, 255, 255)
    print(ColorParser("#000000")); // rgb(0, 0, 0)
    print(ColorParser("#AAA") is Color); // true

    print(ColorParser("rgb(255, 255, 255)")); // rgb(255, 255, 255)
    print(ColorParser("rgba(255, 255, 255, 1.0)")); // rgba(255, 255, 255, 1.0)

    print(ColorParser([10, 20, 30])); // rgb(10, 20, 30)
    print(ColorParser([10, 20, 30, .4])); // rgba(10, 20, 30, 0.4)

    print(ColorParser({"r": 10, "g": 20, "b": 30, "a": .4})); // rgba(10, 20, 30, 0.4)
    print(ColorParser({"x": 10, "y": 20, "z": 30, "a": .4})); // rgba(10, 20, 30, 0.4)

}
```

### List convention

```dart
import 'package:fff/parser.dart';

main() {

    // wrong list of color components
    // but need rgb(10, 20, 30)
    var brg = [30, 10, 20];

    // change list convention
    listConvention = [
      Component.BLUE,
      Component.RED,
      Component.GREEN,
      Component.ALPHA
    ];

    print(ColorParser(brg)); // rgb(10, 20, 30)

}
```

### Map conventions

```dart
import 'package:fff/parser.dart';

main() {

    // by default set [rgba] and [xyza] conventions

    // add new          red   green  blue  alpha
    mapConventions.add(["A",  "B",   "C",  "D"]);

    print(ColorParser({"A": 10, "B": 20, "C": 30, "D": .9})); // rgba(10, 20, 30, 0.9)

    // Attention! Of all the conventions will be applied first best match.
    // Example:

    // remove all conventions
    mapConventions.clear();

    // add new
    mapConventions.add(["a", "b", "c", "A"]);
    mapConventions.add(["x", "y", "z", "A"]);
    mapConventions.add(["1", "2", "3", "A"]);

    // Notice: Need specify all four components (rgba).
    // Convention with the number of components other than the four will be ignored.

    // apply first convention (match "a", "b" and "c")
    print(ColorParser({"a": 10, "b": 20, "c": 30})); // rgb(10, 20, 30)

    // apply second convention (match "x", "y" and "z")
    print(ColorParser({"x": 10, "y": 20, "z": 30})); // rgb(10, 20, 30)

    // apply first convention (math "a" and "b")
    // "c" component is not found
    print(ColorParser({"a": 10, "b": 20, "z": 30})); // rgb(10, 20, 0)

    // apply second convention (math "y" and "z")
    // "x" component is not found
    print(ColorParser({"a": 10, "y": 20, "z": 30})); // rgb(0, 20, 30)

    print(ColorParser({"1": 10, "x": 20, "a": 30})); // rgb(30, 0, 0)

    // if the number of matches in several conventions is identical, then will be applied first

}
```

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/vdakalov/fff/issues


CHANGELOG.md
# Changelog

## 2.0.0

- Add null-safety

## 1.0.1

- Support for dart 2.0.0 SDK

## 1.0.0

- Change self-enum color components with Dart builtin enums
- Remove unused enum items (RGB, RGBA, HEX)
- Move color components from `package:fff/parser.dart` to `package:fff/color.dart`
- Restored tests to work with the new Dart library
- Added the ability to specify a pattern when translating a color into a string

## 0.0.3

- Added method toList for the Color object
- Update docs for constants in presets library
- Added method toMap for the Color object

## 0.0.2+1

- Remove excess annotations types
- Fix changelog version

## 0.0.2

- Simple Color constructor was changed to constant constructor
- Drew parse method in a separate factory ColorParse
- Add 275 preset [Material design](http://www.google.com/design/spec/style/color.html)
- Add more tests

## 0.0.1

- Implements Color class


lib/fff.dart
// Copyright (c) 2015, Viktor Dakalov. All rights reserved. Use of this source code
// is governed by a BSD-style license that can be found in the LICENSE file.

library fff;

_min(num a, num b) => a <= b ? a : b;
_max(num a, num b) => a >= b ? a : b;
_hexin(String comp) => comp.length == 1 ? comp + comp : comp;
_hexout(String comp) => comp.length == 1 ? "0$comp" : comp;
_tohex(num comp) => _hexout(comp.toInt().toRadixString(16));
_fromhex(String comp) => int.parse(_hexin(comp), radix: 16);
_fromnum(String num) => num.contains(".") ? double.parse(num.trim()) : int.parse(num.trim());

List<num> _parseComponents(String input, String prefix, [String postfix = ")"]) {
  if (input.startsWith(prefix) && input.contains(postfix)) {
    return new List<num>.from(
        input.substring(prefix.length, input.indexOf(postfix)).split(",")
        .map(_fromnum).toList());
  }
  return <num>[];
}

List<num> _parseRgb(String input) => _parseComponents(input, "rgb(");

List<num> _parseRgba(String input) => _parseComponents(input, "rgba(");

List<num> _parseHex(String input) {

  if (input.length == 3) {
    return [input.substring(0, 1),
            input.substring(1, 2),
            input.substring(2, 3)].map(_fromhex).toList();
  }

  if (input.length == 6) {
    return [input.substring(0, 2),
            input.substring(2, 4),
            input.substring(4, 6)].map(_fromhex).toList();
  }

  return [];
}

List _parseList(List<num> input) {
  int ri = listConvention.indexOf(RED),
      gi = listConvention.indexOf(GREEN),
      bi = listConvention.indexOf(BLUE),
      ai = listConvention.indexOf(ALPHA);

  return [ri >= 0 && input.length > ri ? input[ri] : null,
          gi >= 0 && input.length > gi ? input[gi] : null,
          bi >= 0 && input.length > bi ? input[bi] : null,
          ai >= 0 && input.length > ai ? input[ai] : null];
}

List _parseMap(Map<dynamic, num> input) {
  var convs = new List<List<String>>.from(mapConventions).reversed.toList(),
      conv,
      result = new List(),
      num = 4;

  if (convs.length == 0) {
    throw new Exception("To create an object of Color from data a map type, you need specify one or more map conventions");
  }

  for (int index = 0; index < convs.length; index++) {
    var conv = convs[index], c = conv.length;

    for (int j = 0; j < conv.length; j++) {
      if (input.containsKey(conv[j])) {
        c--;
      }
    }

    if (c >= result.length) {
      result.length = c + 1;
    }

    result.insert(c, conv);
  }

  if (result.length == 0) {
    throw new Exception("The properties of the map does not match any of the conventions");
  }

  conv = result.firstWhere((c) => c is List);

  return [input[conv[0]], input[conv[1]], input[conv[2]], input[conv[3]]];
}

List _parseArgs(List<num> args) {
  return [args.length > 0 && args[0] is num ? args[0].toInt() : DEF_RED,
          args.length > 1 && args[1] is num ? args[1].toInt() : DEF_GREEN,
          args.length > 2 && args[2] is num ? args[2].toInt() : DEF_BLUE,
          args.length > 3 && args[3] is num ? args[3].toDouble() : DEF_ALPHA];
}

const String RGB = "rgb";
const String RGBA = "rgba";
const String HEX = "hex";

const String RED = "r";
const String GREEN = "g";
const String BLUE = "b";
const String ALPHA = "a";

/// Default red component
num DEF_RED = 0;

/// Default green component
num DEF_GREEN = 0;

/// Default blue component
num DEF_BLUE = 0;

/// Default alpha component
double DEF_ALPHA = null;

/// Agreement on the map color format
List<List<String>> mapConventions = [['r','g','b','a'],['x','y','z','a']];

/// Agreement on the list color format
List<String> listConvention = [RED, GREEN, BLUE, ALPHA];


/// Class for works with color
///
/// from rgb components
///
///     new Color(115, 225, 55);
///     Color.parse("rgb(115, 225, 55)");
///
/// from rgba components
///
///     new Color(90, 90, 255, 0.5);
///     Color.parse("rgba(90, 90, 255, 0.5)");
///
/// from hex string
///
///     new Color.fromHex("FfAacc");
///     Color.parse("#ffaaCC");
///
/// from rgba list
///
///     //          G  B  R  A
///     var gbra = [2, 3, 1, 0];
///     // change convention to [g,b,r,a] (defaul convention [r,g,b,a])
///     Color.listConvention = [Color.GREEN, Color.BLUE, Color.RED, Color.ALPHA];
///     new Color.fromList(grba);
///
/// from map
///
///     var abc = {"AA": 10, "BB": 20, "CC": 30};
///     //                        R     G     B
///     Color.mapConventions.add(["AA", "BB", "CC"]);
///     new Color.fromMap(abc);
///
/// default colors
///
///     Color.black() == new Color(0, 0, 0) == Color.parse('#000000')
///
class Color {

  /// Return red component (0-255).
  final int red;

  /// Return green component (0-255).
  final int green;

  /// Return blue component (0-255).
  final int blue;

  /// Return opacity level (0-1)
  final double alpha;

  /// Constructor create [Color] from rgba components.
  /// Instead not transmitted component will be used by default components
  const Color._(this.red, this.green, this.blue, this.alpha);

  factory Color([dynamic red, num green, num blue, num alpha]) {

    List<num> args = [];

    if (red is String) {
      if (red.contains(")")) {

        if (red.startsWith("rgba(")) {
          args = _parseComponents(red, "rgba(");

        } else if (red.startsWith("rgb(")) {
          args = _parseComponents(red, "rgb(");

        }

      } else if (red.startsWith("#")) {
        args = _parseHex(red.substring(1));

      }

    } else if (red is List<num>) {
      args = _parseList(red);

    } else if (red is Map<dynamic, num>) {
      args = _parseMap(red);

    } else {
      args = [red, green, blue, alpha];

    }

    args = _parseArgs(args);

    return new Color._(args[0], args[1], args[2], args[3]);
  }

  bool operator ==(dynamic other) {
    other = other is Color ? other : new Color(other);
    return red == other.red && green == other.green && blue == other.blue &&
        alpha == other.alpha;
  }

  Color operator +(Color other) {
    return new Color(
        _min(red + other.red, 255),
        _min(green + other.green, 255),
        _min(blue + other.blue, 255),
        alpha != null && other.alpha != null ? _min(alpha + other.alpha, 1) : null);
  }

  Color operator -(Color other) {
    return new Color(
        _max(red - other.red, 0),
        _max(green - other.green, 0),
        _max(blue - other.blue, 0),
        alpha != null && other.alpha != null ? _max(alpha - other.alpha, 0) : null);
  }

  /// Create new [Color] from it
  Color clone() {
    return new Color(red, green, blue, alpha);
  }

  /// Output color as string in format specified [outputFormat]
  String toString() => alpha == null ? toRgbString() : toRgbaString();

  /// Output color in rgba format
  String toRgbaString() => "rgba($red, $green, $blue, ${alpha == null ? 1.0 : alpha})";

  /// Output color in rgb format
  String toRgbString() => "rgb($red, $green, $blue)";

  /// Ouput color in hex format
  String toHexString() => "#${_tohex(red)+_tohex(green)+_tohex(blue)}";

}